llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,281 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+              return false;
+          } else {
+            return false; // CountB is not PHI.
+          }
+        }
+      } else {
+        return false; // CountA is not PHI.
+      }
+    }
+    return true;
+  }
+
+  bool notOverlap(const IntrinsicChainInfo &Other) const {
+    // Different number of updates in the chain.
+    // * but maybe with consistent updates (same steps)
+    if (!hasSameTotalNumberOfCalls(Other))
+      return false;
+
+    // Different parameters.
+    if (!hasSameParameters(Other))
+      return false;
+
+    // Every counter must start equally.
+    if (!hasSameStartingCounters(Other))
+      return false;
+
+    // Check for different steps of calculation.
+    if (hasSameNumberOfCalls(Other))
+      return false;
+
+    return true;
+  }
+};
+
+static IntrinsicArgsInfo collectIntrinsicOperands(const Value *V) {
+  const auto *Extract = dyn_cast<ExtractValueInst>(V);
+  assert(Extract && "Not an Extract");
+  const auto *IntrinsicCall =
+      dyn_cast<CallBase>(Extract->getAggregateOperand());
+  assert(IntrinsicCall && "Not an Intrinsic");
+  Intrinsic::ID ID = IntrinsicCall->getIntrinsicID();
+  IntrinsicArgsInfo Args;
+  unsigned InPtrIdx;
+  isAIEPtrAddIntrinsic(ID, InPtrIdx);
+  // Necessary operands to disambiguate.
+  // Constants.
+  for (unsigned Op : getAddIntrinsicOps(ID)) {
+    Args.ArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  // Counters.
+  for (unsigned Op : getAddIntrinsicCounterOps(ID)) {
+    Args.CounterArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  return Args;
+}
+
+// We can reach a pointer definition starting from a counter.
+static bool
+isAddressingIntrinsicOutput(ArrayRef<const Value *> AdressingOutputs,
+                            const Value *Intrinsic) {
+
+  for (const Value *Counter : AdressingOutputs) {
+    if (Intrinsic != lookThroughAIEAddressingModes(Counter))
+      return false;
+  }
+
+  return true;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *TargetPointer to the Phi node that firstly defines
+/// a pointer value to the same ambiguous memory. We start the search from the
+/// last update of the pointer.
+std::optional<IntrinsicChainInfo>
+trackIntrinsicChain(const Value *LastUpdate, const Value *TargetPointer) {
+  IntrinsicChainInfo ChainInfo;
+  // We count backwards.
+  unsigned LastCalls = 0;
+  // Let's expose the real pointer.
+  TargetPointer = skipCast(TargetPointer);
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    LastUpdate = skipCast(LastUpdate);
+    if (!LastUpdate->getType()->isPointerTy())
+      return std::nullopt; // not a pointer.
+
+    if (auto *GEP = dyn_cast<GEPOperator>(LastUpdate)) {
+      if (!GEP->hasAllZeroIndices())
+        return std::nullopt; // No pointer arithmetic is allowed.
+      // The result pointer and the first operand have
+      // the same value
+      LastUpdate = GEP->getPointerOperand();
```

Comment:
Could we wrap this case in skipCast() ?

Reply:
I think the purpose is different. For example, skipCast always return return a non null pointer, and if the GEP is not a cast (non-zero indexes) we need to handle this case in a different way, like an optional return or a null ckeck after the call. I feel that it can complicate the readability.

Reply:
I think the purpose is to separate the updates from the no-ops. Conceptually, a zero index GEP *is* a pointer cast.

Reply:
Sure, but in this case we will need to handle the GEP twice, considering the null (skipCast)  and non-null (tracking function). As is, we cast just once to look to the offset.